### PR TITLE
Improve efficiency of `vtorc` topo calls 

### PR DESCRIPTION
**Pull request by timvaillancourt**
## Description

This PR adds [a similar optimization to this PR](https://github.com/vitessio/vitess/pull/14693) to `vtorc` by moving to using `ts.GetTabletsByCell(...)` and the new `ts.GetTabletsByShard(...)` methods to fetch tablets. These methods respect the concurrency limit flag for the topo and seem to be more efficient overall

Support for filtering by keyspace/shard was added to `ts.GetTabletsByCell(...)` which is called by `ts.GetTabletsByShard(...)`

[Related to this discussion](https://github.com/vitessio/vitess/issues/16761), we're unfortunately forced to fetch "all" records from the cell just to filter them out by keyspace/shard using `x != y`. In our production when fetching topo records for a single shard, < 0.5% of the topo records fetched are actually used - but that's nothing new

## Related Issue(s)

Resolves https://github.com/vitessio/vitess/issues/17073

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/topo/shard.go
Comment (timvaillancourt) [inline go/vt/topo/shard.go:633]: I copied what other methods do here, but I wonder if this would be more useful after the `ts.GetCellInfoNames(ctx)` fallback `if len(cells) == 0`

File: go/vt/topo/tablet.go
Comment (mattlord) [inline go/vt/topo/tablet.go:270]: Nit, but we can keep the older optimization, using the list size for cap rather than len, while avoiding over allocating in the new path:
```
var sizeHint int
if opt.Keyspace == nil && opt.Shard == nil {
    sizeHint = len(listResults)
}
tablets := make([]*TabletInfo, 0, sizeHint)
```

File: go/vt/topo/tablet.go
Comment (mattlord) [inline go/vt/topo/tablet.go:277]: Nit, but we can combine these.

File: go/vt/topo/tablet.go
Comment (mattlord) [inline go/vt/topo/tablet.go:280]: This means Shard is ignored when Keyspace is not passed. That makes sense, I think. But that also means that I think we can improve the structure. There's already a struct/type that we can use:
```
type KeyspaceShard struct {
	Keyspace string
	Shard    string
}
```

We can add a member of that type to `GetTabletsByCellOptions` instead. Then we can also simplify the hint note above to:
```
var sizeHint int
if opt.KeyspaceShard == nil {
    sizeHint = len(listResults)
}
tablets := make([]*TabletInfo, 0, sizeHint)
```

**Review by mattlord — COMMENTED** [review]
This is looking good! ❤️ I only had some minor nits and suggestions. Let me know what you think.

File: go/vt/topo/tablet.go
Comment (mattlord) [reply go/vt/topo/tablet.go:270]: Oh, and opt is a pointer. So we should check for nil when looking at its members as well.

**Review by GuptaManan100 — APPROVED** [review]
Apart from Matt's comments, everything else LGTM. It can be merged once they have been addressed.

File: go/vt/topo/shard.go
Comment (mattlord) [reply go/vt/topo/shard.go:659]: Alternatively we could use the literal 3 as the allocation hint / initial capacity.

File: go/vt/topo/shard.go
Comment (timvaillancourt) [reply go/vt/topo/shard.go]: Updated

File: go/vt/topo/shard.go
Comment (timvaillancourt) [reply go/vt/topo/shard.go]: Updated

File: go/vt/topo/shard.go
Comment (timvaillancourt) [reply go/vt/topo/shard.go]: Updated. Also updated `FindAllTabletAliasesInShardByCell` I copied this from

File: go/vt/topo/tablet.go
Comment (timvaillancourt) [reply go/vt/topo/tablet.go]: Updated

File: go/vt/topo/tablet.go
Comment (timvaillancourt) [reply go/vt/topo/tablet.go]: Nice! Updated

File: go/vt/topo/tablet.go
Comment (timvaillancourt) [reply go/vt/topo/tablet.go]: Updated

File: go/vt/topo/shard.go
Comment (timvaillancourt) [reply go/vt/topo/shard.go]: Updated

**Comment by timvaillancourt** [comment]
Thanks for the review! Updated with @mattlord's suggestions 👍

File: go/vt/topo/shard.go
Comment (mattlord) [inline go/vt/topo/shard.go:670]: Nit, but now that we're doing this in a function we can unlock in a defer:
```
defer mu.Unlock()
```

**Comment by timvaillancourt** [comment]
Whoops, meant to add ^^^ backport labels to a different PR. Removed

File: go/vt/topo/tablet.go
Comment (mattlord) [inline go/vt/topo/tablet.go:240]:
```suggestion
	// An empty shard value will match all shards in the keyspace.
```

File: go/vt/topo/tablet_test.go
Comment (mattlord) [inline go/vt/topo/tablet_test.go:77]: Probably worth adding another case where the shard is empty, confirming that we match all shards in that Keyspace.

File: go/vt/topo/tablet.go
Comment (mattlord) [inline go/vt/topo/tablet.go:286]: So you can pass in an empty keyspace but a shard of "-" e.g. to match tablets in any unsharded keyspace. Is that intentional? I'm not sure it's wrong, but checking as the test currently assumes there's only 1 matching keyspace but here there could be N.

File: go/vt/topo/tablet_test.go
Comment (mattlord) [inline go/vt/topo/tablet_test.go:84]: IMO it's worth making this more explicit in the test. For example:
```diff
diff --git a/go/vt/topo/tablet_test.go b/go/vt/topo/tablet_test.go
index c61e806579..5ac7d05d39 100644
--- a/go/vt/topo/tablet_test.go
+++ b/go/vt/topo/tablet_test.go
@@ -19,6 +19,7 @@ package topo_test
 import (
        "context"
        "errors"
+       "fmt"
        "testing"

        "github.com/stretchr/testify/assert"
@@ -145,10 +146,18 @@ func TestServerGetTabletsByCell(t *testing.T) {
                                require.Equal(t, tab.Tablet, tablets[i].Tablet)
                        }

+                       shards := make(map[string]struct{}, len(tt.keyspaceShards))
                        for _, tablet := range out {
+                               shards[fmt.Sprintf("%s/%s", tablet.Keyspace, tablet.Shard)] = struct{}{}
                                require.Equal(t, keyspace, tablet.Keyspace)
                                require.Equal(t, shard, tablet.Shard)
                        }
+                       if tt.opt != nil && tt.opt.KeyspaceShard != nil {
+                               t.Logf("Filtering shards. Total shards: %d, matching shards: %d", len(tt.keyspaceShards), len(shards))
+                               require.Len(t, shards, 1)
+                       } else {
+                               require.Equal(t, len(tt.keyspaceShards), len(shards))
+                       }
                })
        }
 }
```

Comment (mattlord) [reply go/vt/topo/tablet_test.go:84]: That's assuming, however, that we don't support passing in an empty keyspace with a non empty shard though.

**Review by mattlord — COMMENTED** [review]
We're super close! I really only had the one open question about shard filtering when there's no Keyspace provided.

File: go/vt/topo/tablet_test.go
Comment (mattlord) [reply go/vt/topo/tablet_test.go:84]: Either way, probably even better to have a test case member that explicitly contains the expected matches. Then that can be verified, not only the count but that we got the expected results. Right now we're not actually testing and confirming the results when filtering is done.

Comment (mattlord) [reply go/vt/topo/tablet_test.go:84]: Here's an example of what I mean above:
```diff
diff --git a/go/vt/topo/tablet_test.go b/go/vt/topo/tablet_test.go
index c61e806579..4afcb625b4 100644
--- a/go/vt/topo/tablet_test.go
+++ b/go/vt/topo/tablet_test.go
@@ -44,6 +44,7 @@ func TestServerGetTabletsByCell(t *testing.T) {
 		opt                *topo.GetTabletsByCellOptions
 		listError          error
 		keyspaceShards     map[string]string
+		expectedShards     map[string]string
 	}{
 		{
 			name:               "negative concurrency",
@@ -90,6 +91,7 @@ func TestServerGetTabletsByCell(t *testing.T) {
 					Shard:    shard,
 				},
 			},
+			expectedShards: map[string]string{keyspace: shard},
 		},
 	}
 
@@ -145,10 +147,17 @@ func TestServerGetTabletsByCell(t *testing.T) {
 				require.Equal(t, tab.Tablet, tablets[i].Tablet)
 			}
 
+			shards := make(map[string]string, len(tt.keyspaceShards))
 			for _, tablet := range out {
+				shards[tablet.Keyspace] = tablet.Shard
 				require.Equal(t, keyspace, tablet.Keyspace)
 				require.Equal(t, shard, tablet.Shard)
 			}
+
+			if tt.opt != nil && tt.opt.KeyspaceShard != nil {
+				require.NotEmpty(t, tt.expectedShards)
+				require.Equal(t, tt.expectedShards, shards)
+			}
 		})
 	}
 }

```

File: go/vt/topo/shard.go
Comment (mattlord) [inline go/vt/topo/shard.go:663]: I think the common path will have keyspace and shard both as empty strings? If so, we could pass a nil KeyspaceShard in those cases to make the work done in GetTabletsByCell a little easier:
```
	tablets := make([]*TabletInfo, 0, len(cells))
	var kss *KeyspaceShard
	if keyspace != "" || shard != "" {
		kss = &KeyspaceShard{
			Keyspace: keyspace,
			Shard:    shard,
		}
	}
	options := &GetTabletsByCellOptions{
		Concurrency:   cellConcurrency,
		KeyspaceShard: kss,
	}
```

File: go/vt/topo/tablet.go
Comment (mattlord) [reply go/vt/topo/tablet.go:286]: If we do want to support that (shard w/o a keyspace), let's add a test case for it.

File: go/vt/topo/tablet_test.go
Comment (mattlord) [inline go/vt/topo/tablet_test.go:53]: All of these `[]string` specifications are redundant given the defined type for the values in the map. Doesn't hurt anything though either, but it makes things a little tidier:
```
keyspace: {shard},
```

**Review by mattlord — COMMENTED** [review]
Looking good! The only remaining issue is whether or not we want to support filtering by shard w/o a keyspace. I'm OK either way, but if we do want to support that then we should add a test case for it. That make sense?

File: go/vt/topo/tablet_test.go
Comment (mattlord) [reply go/vt/topo/tablet_test.go:53]: I think this is why gofmt is failing:
```
Run $(go env GOPATH)/bin/golangci-lint run go/... --timeout 10m || exit 1
  $(go env GOPATH)/bin/golangci-lint run go/... --timeout 10m || exit 1
  shell: /usr/bin/bash -e {0}
go/vt/topo/tablet_test.go:53: File is not `gofmt`-ed with `-s` (gofmt)
				keyspace: []string{shard},
```

Comment (timvaillancourt) [reply go/vt/topo/tablet_test.go]: Good idea, updated

File: go/vt/topo/shard.go
Comment (timvaillancourt) [reply go/vt/topo/shard.go]: Good idea, updated

**Comment by timvaillancourt** [comment]
> The only remaining issue is whether or not we want to support filtering by shard w/o a keyspace

@mattlord I don't plan to support this because I can't think of a use case. I've updated to make keyspace required 👍

File: go/vt/topo/tablet.go
Comment (timvaillancourt) [reply go/vt/topo/tablet.go]: Updated to require a keyspace

File: go/vt/topo/tablet.go
Comment (timvaillancourt) [inline go/vt/topo/tablet.go:292]: @mattlord I've converted the test to check exact tablets as suggested. A consequence of this is we need to sort the output

I can make this conditional for testing but I see we sort results in other places that do concurrent gets, so I'll hold off for now

Comment (timvaillancourt) [reply go/vt/topo/tablet.go:292]: I suppose the sort could move to `tablet_test.go` if there is a concern about cost

File: go/vt/topo/tablet.go
Comment (mattlord) [inline go/vt/topo/tablet.go]: Is this really just for tests (predictable and consistent results)? If so, we could order the results there in the test instead. Not that I think having predictable results here is a bad thing — I'm just not sure that it's helpful and warrants the cost at this layer. My gut feeling is that the caller should order the entire results that it gets for all tablets in all keyspaces and cells (whatever it's requesting) and then order the full set as it prefers (based on keyspace, cell, UID etc.).

File: go/vt/topo/tablet.go
Comment (timvaillancourt) [reply go/vt/topo/tablet.go]: Updated: moved it to `tablet_test.go` until there's a need for a sort in the normal code

File: go/vt/topo/tablet.go
Comment (mattlord) [reply go/vt/topo/tablet.go]: jinx 🤣

File: go/vt/topo/tablet.go
Comment (mattlord) [reply go/vt/topo/tablet.go:281]: Ah, never mind. This is so that we ignore the Shard value if a Keyspace is not provided.

**Comment by mattlord** [comment]
Looks like the new unit test is flaky: 
  - https://github.com/vitessio/vitess/actions/runs/11691174276/attempts/1?pr=17071
  - https://github.com/vitessio/vitess/actions/runs/11691174293/attempts/1?pr=17071

That's the only thing left! Thank you for sticking this out, @timvaillancourt ! This will be a nice improvement.

File: go/vt/topo/tablet.go
Comment (timvaillancourt) [reply go/vt/topo/tablet.go:281]: I guess is uses less lines but I think a 3 `&&` if condition is already getting hard to read. I generally stop at 2~ conditions and nest an if for readability

File: go/vt/topo/tablet.go
Comment (timvaillancourt) [reply go/vt/topo/tablet.go]: Sounds good. It's moved to the test only now

**Comment by timvaillancourt** [comment]
@mattlord I believe moving `tt.keyspaceShards` to `keyspaceShards []*topo.KeyspaceShard` _(from a `map`)_ will fix this flakiness 👍

**Review by mattlord — APPROVED** [review]
Great work on this, @timvaillancourt ! You can merge any time, just please don't forget to squash and merge. 

One last note, up to you if you want to add it:
```

			slices.SortFunc(out, func(i, j *topo.TabletInfo) int {
				return cmp.Compare(i.Alias.Uid, j.Alias.Uid)
			})
			for i, tablet := range out {
				expected := tt.expectedTablets[i]
				require.Equal(t, expected.Alias.String(), tablet.Alias.String())
				require.Equal(t, expected.Keyspace, tablet.Keyspace)
				require.Equal(t, expected.Shard, tablet.Shard)
			}
```

Right now you're ordering all of entries in `expectedTablets` by Uid. IMO it's worth sorting that slice as well so that things do not become flaky in the future if someone modifies the existing test cases or adds new ones where that slice is not ordered:
```

			slices.SortFunc(out, func(i, j *topo.TabletInfo) int {
				return cmp.Compare(i.Alias.Uid, j.Alias.Uid)
			})
			slices.SortFunc(tt.expectedTablets, func(i, j *topo.TabletInfo) int {
				return cmp.Compare(i.Alias.Uid, j.Alias.Uid)
			})
			for i, tablet := range out {
				expected := tt.expectedTablets[i]
				require.Equal(t, expected.Alias.String(), tablet.Alias.String())
				require.Equal(t, expected.Keyspace, tablet.Keyspace)
				require.Equal(t, expected.Shard, tablet.Shard)
			}
```

**Comment by timvaillancourt** [comment]
@mattlord thanks for the reviews, added that!

**Pull request merged**
